idl/group0_tables_lang.idl.hh

```diff
@@ -0,0 +1,27 @@
+/*
+ * Copyright 2022-present ScyllaDB
+ */
+
+/*
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+
+namespace raft {
+namespace group0_tables {
+
+struct select_query {
+    bytes key;
+};
+
+struct update_query {
+    bytes key;
+    bytes new_value;
+    std::optional<bytes_opt> value_condition;
+};
+
+struct query {
+    std::variant<select_query, update_query> q;
+};
+
```

Comment:
How will we scale this to encompass all of CQL?

Reply:
Actually I think this should follow the internal C++ api both in language and supported operations, otherwise we might end up marrying this to a single front-end language.

Reply:
@denesb what do you mean by "the internal C++ api"?

Reply:
> How will we scale this to encompass all of CQL?

Answering this question is a big task which will have its own design and implementation stages.

Reply:
I think we should use CQL and prepared statements. We already have a huge mechanism to represent CQL (CQL language), deserialize it (the parser), cache the parsing work (prepared statements cache) and manage a remote handle the the cache (prepared statement ID).

Reply:
1. If we want to use prepared statements cache, we cannot put the CQL string inside the Raft command - that would require every replica to parse it on apply.
2. But we also cannot put a prepared statement ID there, because we have no guarantee that the cache is the same on every replica.

Maybe we could put something like (string, ID) pair, a replica would first try to fetch the statement from its local cache using the ID, if it's not present then parse the string.
But to make this work efficiently we would need to ensure that the function CQL string -> ID is deterministic.

Reply:
But I also don't want to put anything CQL specific inside the Raft commands because I want to make it possible to support other frontends.

We'll probably want to use the storage_proxy data structures in the end: `read_command`, `mutation` and so on. Both CQL and Alternator use them so at least those frontends would be covered.

Reply:
The prepared statement ID is a hash of the CQL string, so that part isn't a problem.

There is a problem if the recipient doesn't have the CQL string, but it can just respond with a rejection, and the sender can retry with the CQL string included.

Reply:
The other frontends can standardize on CQL as their language. Is CQL the perfect language? No. Is the language you're sketching on a napkin the perfect language? Also no.

Reply:
> The other frontends can standardize on CQL as their language.

@nyh or @psarna could you give your input here? Would it be possible to compile Alternator to CQL?

Reply:
> Is the language you're sketching on a napkin the perfect language? Also no.

This language is internal and we can change it as we like it or need it. CQL, not so much.

Also, I'm proposing we try to reuse the data structures from `storage_proxy`.

Reply:
The piece of code we're currently discussing is the definition of the data structure which is put inside a Raft command, the thing that is replicated to every replica. Since you commented on it, I understood that you want to put the whole CQL statement inside (or equivalent of it).

Reply:
> Then it generates a read_command for reads or a mutation for writes. Replicas don't need the full statement.

So you're basically agreeing that we want to use the `storage_proxy` data structures (`read_command`, `mutation`) for the Raft commands.

Now the question remains, what do we put there for conditional statements.

Reply:
> @denesb what do you mean by "the internal C++ api"?

Basically the `storage_proxy` API for reads and writes: `query()` and `mutate()`.

Reply:
> So you're basically agreeing that we want to use the `storage_proxy` data structures (`read_command`, `mutation`) for the Raft commands.

This is basically what I meant by our "C++ api".

> Now the question remains, what do we put there for conditional statements.

A conditional statement is basically an additional read (`read_command`) and a condition expression.

Reply:
> > The other frontends can standardize on CQL as their language.
> 
> @nyh or @psarna could you give your input here? Would it be possible to compile Alternator to CQL?

Compiling to a legal CQL query which can be parsed and processed by Scylla, no, not trivially - otherwise we wouldn't bother with coding the whole frontend layer and we'd just translate everything as is to CQL. But, especially with the new expression-based system, it's most likely possible to translate alternator requests to something that looks like CQL and back. But then, if you're considering serializing requests as strings, why not just store alternator requests, which are JSON, ergo strings? We do not allow accessing the same table from both alternator and CQL, so you might just as well store alternator requests as is.

Reply:
True, we could forward alternator requests.

A downside is that is changes alternator from being a front-end to being involved in mid-layer processing.

Reply:
I had some ideas about recovering prepared statements from non-prepared CQL and alternator requests:

1. Parse CQL/JSON and remember the locations of literals
2. Hash the input text, excluding the literals
3. If the the same hash recurs, create a prepared statement out of the input-text-excluding-literals-with-bind-markers-added (CQL) or a translation of the query to CQL (alternator)
4. add the text-excluding-literals to a recognition automaton (e.g. regex) to recognize it and capture the literals
5. pass queries first via the automaton to see if the match such an extracted prepared statement


The value for CQL is dubious since it encourages bad behavior.

But for alternator, if parsing is sufficiently expensive, perhaps worthwhile.

Reply:
We discussed this and put it off for later when we decide how the command will look like.